File: patterns/2020_10_07/pythonImpl/pair.py
import sys
import time
from datetime import datetime
from pynng import Pair0, Timeout


def send_recv(sock: Pair0, name: str):
    while True:
        try:
            msg = sock.recv()
            print(f'{name}: RECEIVED "{msg.decode()}"')
        except Timeout:
            pass

        time.sleep(1)

        try:
            print(f'{name}: SENDING "{name}"')
            sock.send(name.encode())
        except Timeout:
            pass


def node0(url: str):
    with Pair0(listen=url, recv_timeout=100, send_timeout=100) as sock:
        send_recv(sock, 'NODE0')


def node1(url: str):
    with Pair0(dial=url, recv_timeout=100, send_timeout=100) as sock:
        send_recv(sock, 'NODE1')


if __name__ == "__main__":
    # print(sys.argv)
    if len(sys.argv) > 2 and sys.argv[1] == 'node0':
        node0(sys.argv[2])
    elif len(sys.argv) > 2 and sys.argv[1] == 'node1':
        node1(sys.argv[2])
    else:
        print(f"Usage: {sys.argv[0]} node0|node1 <URL> ...")
        sys.exit(1)

File: patterns/2020_10_07/goImpl/pair/pair.go
package main

import (
	"fmt"
	"os"
	"time"

	"go.nanomsg.org/mangos/v3"
	"go.nanomsg.org/mangos/v3/protocol/pair"

	// register transports
	_ "go.nanomsg.org/mangos/v3/transport/all"
)

func die(format string, v ...interface{}) {
	fmt.Fprintln(os.Stderr, fmt.Sprintf(format, v...))
	os.Exit(1)
}

func sendName(sock mangos.Socket, name string) {
	fmt.Printf("%s: SENDING \"%s\"\n", name, name)
	if err := sock.Send([]byte(name)); err != nil {
		die("failed sending:%s", err)
	}
}

func recvName(sock mangos.Socket, name string) {
	var msg []byte
	var err error
	if msg, err = sock.Recv(); err == nil {
		fmt.Printf("%s: RECEIVED: \"%s\"\n", name, string(msg))
	}
}

func sendRecv(sock mangos.Socket, name string) {
	for {
		sock.SetOption(mangos.OptionRecvDeadline, 100*time.Millisecond)
		recvName(sock, name)
		time.Sleep(time.Second)
		sendName(sock, name)
	}
}

func node0(url string) {
	var sock mangos.Socket
	var err error
	if sock, err = pair.NewSocket(); err != nil {
		die("can't get new pair socket: %s", err)
	}
	if err = sock.Listen(url); err != nil {
		die("can't listen on pair socket:%s", err.Error())
	}
	sendRecv(sock, "node0")
}

func node1(url string) {
	var sock mangos.Socket
	var err error
	if sock, err = pair.NewSocket(); err != nil {
		die("can't get new pair socket: %s", err)
	}
	if err = sock.Dial(url); err != nil {
		die("can't dial on pair socket:%s", err.Error())
	}
	sendRecv(sock, "node1")
}

func main() {
	if len(os.Args) > 2 && os.Args[1] == "node0" {
		node0(os.Args[2])
		os.Exit(0)
	}
	if len(os.Args) > 2 && os.Args[1] == "node1" {
		node1(os.Args[2])
		os.Exit(0)
	}

	fmt.Fprintf(os.Stderr,
		"Usage: pair node0|node1 <URL> <ARG> ...\n")
	os.Exit(1)
}
